`WatchTask` and `StopTask` for AWS ECS Task Launcher

paladin-devops opened the pull request:
Closes #3713.

The task launcher plugin for AWS ECS will use the ECS and CloudWatch AWS APIs to get the logs of the running on-demand runner task in ECS.

paladin-devops committed:
ECS task launcher WatchTask implementation.

paladin-devops committed:
ECS task launcher updates.

Implement ECS StopTask, to kill ODR if StopTask needs to. Simplified concurrency of WatchTask for ECS. Status is being sent over a string channel and used in the main goroutine.

paladin-devops committed:
Comment corrections in k8s task launcher plugin.

paladin-devops committed:
Package rearrangement, comment update, changelog.

paladin-devops committed:
Fix status return on error for number of ECS tasks.

izaaklauer commented on builtin/aws/ecs/task.go:
```diff
@@ -209,24 +212,152 @@ func (p *TaskLauncher) Config() (interface{}, error) {
+		taskResp, err := ecsSvc.DescribeTasks(&ecs.DescribeTasksInput{
+			Cluster: aws.String(p.config.Cluster),
+			Tasks:   aws.StringSlice([]string{*taskArn}),
+			Include: aws.StringSlice([]string{"TAGS"}),
```
oh aws apis.

briancain commented on builtin/aws/ecs/task.go:
```diff
@@ -209,24 +212,152 @@ func (p *TaskLauncher) Config() (interface{}, error) {
+
+	// this channel will receive the status of the ECS task
+	taskStatusCh := make(chan string)
+	go taskStatus(ctx, time.Duration(5*time.Minute), &taskStatusCh, log, sess, p.config.Cluster, ti.Id)
```
Maybe it makes sense to have `taskStatus` with a pointer recevier (i.e. `func (p *TaskLauncher) taskStatus (...)`). Then we could reduce some of these parameters?

briancain commented on builtin/aws/ecs/task.go:
```diff
@@ -209,24 +212,152 @@ func (p *TaskLauncher) Config() (interface{}, error) {
+
+	token := ""
+	var resp *cloudwatchlogs.GetLogEventsOutput
+	for {
```
Is it possible to refactor this to also have a `for { select { ... } }` pattern like the previous loop? I feel like with unbound loops those are really nice for readability 🤔

paladin-devops replied:
@briancain added! But do you also mean to add another timeout context for the `select` to read from?

briancain commented on builtin/aws/ecs/task.go:
```diff
@@ -362,4 +499,60 @@ func roleArn(name string, sess *session.Session) (string, error) {
+		case <-taskContext.Done():
+			log.Error("timeout waiting for task to become ready", terminal.WithErrorStyle())
+			return
+		default:
```
Since the default is essentially "execute the loop", maybe we only have 1 select case for when we timeout, then outside of the select, do this logic?

paladin-devops committed:
Move AWS WatchTask log retrieval into select block.

paladin-devops committed:
Move "default" logic of waiting for the ECS task to start the ODR for WatchTask out of select block.

paladin-devops committed:
Simplify params for ECS task status check function.

paladin-devops committed:
Correctly get cluster from platform config in ECS task status check.

paladin-devops committed:
Make ODR task status a trace level log for ECS.

Reporting the ODR task status repeatedly in the loop makes for noisy logs even at the debug level.

paladin-devops committed:
Remove unnecessary select statement.
